CenterForOpenScience/osf.io
# [IN-167][OSF] Add `/subscriptions` endpoint for managing notification subscriptions for PreprintProviders

## Purpose

This PR enables reading and updating subscription from the API.

## Changes

1. Added ForeignKey relationship to `AbstractProvider` for `NotificationSubscription` model.
2. Added routing for `/subscriptions` endpoint
3. Added routing for `/subscriptions/<subscription_id>` endpoint
4. Added a `SubscriptionList` view for the route
5. Added a `SubscriptionDetail` view for the route.
6. Added a `SubscriptionListSerializer` for the view
7. Added a `SubscriptionDetailSerializer` for the view
8. Added a `FrequencyField` to compute frequency information according to m2m relationships of the `NotificationSubscription` instance.
9. Added a simple permission class `IsSubscriptionOwner` for permission control.
10. In `osf/models/mixins`, logic is added to add/removed users from subscriptions upon adding/removing a user from moderator groups.
11. In `osf/models/providers`, a `post_save` signal handler is added to create `NotificationSubscription` instances for newly created providers.
12. A custom migration (`0084_populate_notification_subscriptions`) is added to create instances of `NotificationSubscription` for existing preprint providers and adding their moderators and admins as subscribers.
13. Changes are made to the admin app so that when OSF admin add users to review groups it would also subscribe them to the preprint provider.

## QA Notes

Test to make sure the existing PreprintProvider's moderators can access their subscription information, which should be "instant" by default. Also, make sure no one can access other people's subscription status through the API.
~More thorough tests (e.g. PATCH/PUT) could be done when UI is ready.~ Alex notes: Please test patch/put now too. Ideally we won't have to modify this to get the front end working.


## Side Effects

None

## Ticket

https://openscience.atlassian.net/browse/IN-167
